### ISSUE TITLE
Question : OpenWriteAsync vs WriteAsync

### ISSUE THREAD

**Issue by willignicolas** [aloneguid/storage#189]
Hello Ivan,

Thanks for your exelent work on storage.net. Our Company use it very often.

In a projet we use a old version of the nuget and the OpenWriteAsync Method on many spot to create a file on the storage and get a stream on the file. After that we write on the stream to ensure to not have all stream on memory.

We have try to migrate the nuget version (to have a correction of a known bug) but the openwriteasync method is no longer available. We can see a WriteAsync method but with this we must pass a stream in parameter.

Have you a method to get a stream from a blob to write on it?

Thanks for your help.

Nicolas.

**Comment by willignicolas** [comment]
Ivan, 

To help you anderstand our situation there is a exemple of OpenWriteAsync() in our code with the old version of the nuget 

` 
using (var archiveStream = await fileProvider.OpenWriteAsync("test.zip"))
            {
                //Création de l'archive
                using (ZipArchive archive = new ZipArchive(archiveStream, ZipArchiveMode.Create, false))
                {

                    //Récupération de la liste des fichiers présents dans le dossier temporaire
                    var fileList = (await _temporaryFiles.ListFiles(TempDirectory, true)).Where(f => f.IsFile).ToList();

                    fileList.ForEach(f =>
                    {
                        //En enlève le dossier temporaire du path
                        var filenameWithoutFirstFolder = String.Join("/", f.FullPath.Split('/').Skip(1));

                        //Création d'une entrée dans l'archive
                        var zipEntry = archive.CreateEntry(filenameWithoutFirstFolder);

                        //Ouverture d'un stream sur un fihcier temporaire
                        using (var sourceFileStream = _temporaryFiles.ReadFile(f.FullPath).Result)
                        {
                            //Ouverture d'un stream sur la nouvelle entrée de l'archive
                            using (Stream entryStream = zipEntry.Open())
                            {
                                //On se place au début du stream
                                sourceFileStream.Position = 0;
                                //Copie du fichier temporaire dans l'entrée de l'archive
                                sourceFileStream.CopyTo(entryStream);
                            }
                        }
                    });
                }
            }
`

We don't anderstand how to use new method Write without hace a stream in our memory.

Thanks,
Nicolas.

**Comment by willignicolas** [comment]
no idea @aloneguid ?

**Comment by aloneguid** [comment]
Hi, yes OprnWriteAsync was removed as it's very inefficient with cloud providers. For instance S3 needs to know file size beforehand, otherwise all the writes need to be kept in memory and that becomes extremely memory intensive. I think only local disk can do it right. You could replace call to OWA with the one from System.IO as a solution?

**Comment by willignicolas** [comment]
> Hi, yes OprnWriteAsync was removed as it's very inefficient with cloud providers. For instance S3 needs to know file size beforehand, otherwise all the writes need to be kept in memory and that becomes extremely memory intensive. I think only local disk can do it right. You could replace call to OWA with the one from System.IO as a solution?

Thanks for your reply @aloneguid,

The great interest to use your Storage library is to avoid calling System.IO (because our application work on premise but is intended to work in a cloud host in the futur)

So in my exemple (to zip files) I really don't know how to avoid extreme memory using. Do you have any lead with using the actual nuget version (for local disk and Azure) ?

**Comment by aloneguid** [comment]
Hi, I'm looking at some options at the moment to support this, and at the same time efficiently support all providers.

**Comment by willignicolas** [comment]
Wow @aloneguid that's great news.

Thanks to take time for this.

Nicolas.

**Comment by willignicolas** [comment]
Hi @aloneguid,

Any news about that ? 

Best regards.

**Comment by jussimattila** [comment]
We are also very interested in getting the old method back. We are using Storage.Net for file operations that are run either against local disk or Azure blob storage, depending on the deployment. The common interface for reading and writing files/blobs was one of the primary reasons for choosing this library.

**Comment by aloneguid** [comment]
@jussimattila holding an open write stream is simply not possible outside of local filesystem, and the dirty trick that used to be implemented before was leading to memory issues under server pressure. Azure SDK does it in a really workaround way that makes you to pay more for storage and potentially undeleted data stuck on the backend. If you still want to have it you could create an extension method `OpenWriteAsync` that returns a new memorystream and dumps data to cloud on Dispose, it will be similar efficiency.

**Comment by willignicolas** [comment]
@aloneguid @jussimattila The point of having IBlobStorage.OpenReadAsync()/IBlobStorage.OpenWriteAsync() is to not have big file on memory.

**Comment by jussimattila** [comment]
I understand the limitations. We previously used an in-house abstraction layer for enabling switching between local/cloud storage. We then switched to Storage.Net, as it did everything ours did plus more. We will definitely implement an extension method if you decide blob write streams will not be supported, and I can also understand the reasoning. It's just that the previous interface was perfect for our use case, even with the limitations involved.

**Comment by jussimattila** [comment]
I also agree with @willignicolas that the interface is needed to be able to stream large files to remote storage. For azure, see this example: https://www.andrewhoefling.com/Blog/Post/uploading-large-files-to-azure-blob-storage-in-c-sharp

**Comment by aloneguid** [comment]
@willignicolas you did have big files in memory, it was just handled in SDK external to your code :( The problem is library users will blame the library for inefficient memory usage.

@jussimattila the sample you've provided has other issues, on the surface:
- you are paying at least twice more for azure storage calls, as it needs to have minimum 2 calls - upload chunk and commit, instead of one.
- it doesn't actually work as a writeable stream, you need to provide your own stream as a source.
- The uploads are slow, unless source stream is seekable.

And this is exactly how native Azure Blob SDK works internally - when you call OpenWriteAsync the following happens:
1) it checks blob exists (+1 rest call)
2) create a virtual stream and immediately returns
3) when you write to the stream, if there is enough data, that stream periodically dumps the buffer to cloud storage (+1 rest call per buffer)
4) on dispose, sdk commits all the chunks (+1 rest call). If you forget to call dispose, the uploaded chunks will still be in Azure and you will be billed for them until they are cleared up on schedule by azure itself - there is no way to see them or delete manually.

In addition to this, nothing happens in parallel for large files. Other clouds (AWS, GCP) don't even attempt to support this scenario as it's very inefficient.

So yeah, we can support OpenWriteAsync, but it will only work properly for local filesystem and should be avoided for anything else. Cloud storage does not support efficient uploads of unknown file sizes, period.
